define/types.go

```diff
@@ -191,16 +202,54 @@ func cloneToDirectory(url, dir string) ([]byte, string, error) {
 		}
 		gitBranch = gitSubDirPart[0]
 	}
+	// Attempt parsing commit, accepted syntax listed below:
+	// * repourl.git#branch:subdir?commit
+	// * repourl.git#branch?commit ( uses default subdir )
+	// * repourl.git?commit ( uses default branch )
+	if len(gitBranchPart) > 1 {
+		if gitSubdir != "" {
+			gitCommitPart := strings.Split(gitSubdir, "?")
+			if len(gitCommitPart) > 1 {
+				gitCommit = gitCommitPart[1]
+			}
+			gitSubdir = gitCommitPart[0]
+		} else {
+			gitSubDirPart := strings.Split(gitBranchPart[1], "?")
+			if len(gitSubDirPart) > 1 {
+				gitCommit = gitSubDirPart[1]
+			}
+			gitBranch = gitSubDirPart[0]
+		}
+	}
+	return gitBranchPart[0], gitSubdir, gitCommit, gitBranch
+}
+
+func cloneToDirectory(url, dir string) ([]byte, string, error) {
+	var cmd *exec.Cmd
+	gitRepo, gitSubdir, gitCommit, gitBranch := ParseGitBuildContext(url)
 	if gitBranch == "" {
-		logrus.Debugf("cloning %q to %q", gitBranchPart[0], dir)
-		cmd = exec.Command("git", "clone", "--recurse-submodules", gitBranchPart[0], dir)
+		logrus.Debugf("cloning %q to %q", gitRepo, dir)
+		cmd = exec.Command("git", "clone", "--recurse-submodules", gitRepo, dir)
 	} else {
-		logrus.Debugf("cloning repo %q and branch %q to %q", gitBranchPart[0], gitBranch, dir)
-		cmd = exec.Command("git", "clone", "--recurse-submodules", "-b", gitBranch, gitBranchPart[0], dir)
+		logrus.Debugf("cloning repo %q and branch %q to %q", gitRepo, gitBranch, dir)
+		cmd = exec.Command("git", "clone", "--recurse-submodules", "-b", gitBranch, gitRepo, dir)
```

Comment:
When we do builds we typically only need the files from a specific commit in the git repo. Could we re-work this at some point to pull only a specific commit (--depth=0) for efficiency sake?

Reply:
Sure we could do this If this becomes bottleneck in future just kept it similar to branch pulling for consistency.

Reply:
Thanks! Should I open a feature request?

The reason I think it's so important is that some repos have a really long history and a bunch of vendored deps included. It would be a great improvement to do shallow pulls.

As an example building the coreos-assembler repo if I pull the entire history is like 70M whereas if I just pull a single commit it's like 7M. That's quite a difference.

Reply:
oh that's great!

To be clear I think we should be able to make this optimization for all builds from a git repo. Grab me in `#podman` on IRC if you want to discuss further!

Reply:
@nalind as @dustymabe suggested could we change `--depth=1` for all use-cases, I am unable to think of any reason that is blocker for this request but just want to confirm if there is any hidden edge case.

Reply:
Note that supporting `--depth=1` is a bit trickier when someone specifies a commit. i.e. you can't just `git clone`, but you can support it doing something like what I show in https://github.com/containers/buildah/issues/4148#issue-1322488191 with `git init; git fetch, etc..`

Reply:
Implemented in latest commit.